task8/ServletsOnMaven/src/main/java/WorkWithDB/WorkWithDB.java

```diff
@@ -0,0 +1,316 @@
+            } else {
+                PreparedStatement statementToInsertLike = conn.prepareStatement(
+                        "insert likes(idOfPost, username) value (?, ?)");
+                statementToInsertLike.setString(1, idOfPost);
+                statementToInsertLike.setString(2, username);
+                statementToInsertLike.executeUpdate();
+                return 1;
+            }
+
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+        return 0;
+    }
+
+    public static Post getPostById(String id, DataSource dataSource) {
+        try (Connection conn = dataSource.getConnection()) {
+
+            PreparedStatement statementToGetPostById = conn.prepareStatement(
+                    "select * from posts where idOfPost=?"
+            );
+            statementToGetPostById.setString(1, id);
+            ResultSet result = statementToGetPostById.executeQuery();
+            if (!result.next()) {
+                return null;
+            }
+            Post post = new Post(id,
+                    result.getString("username"),
+                    LocalDateTime.parse(result.getString("createdAt").replace(' ', 'T')),
+                    result.getString("description"),
+                    new ArrayList<>(),
+                    new ArrayList<>());
+
+            PreparedStatement statementToGetLikes = conn.prepareStatement(
+                    "select username from likes where idOfPost = ?");
+            statementToGetLikes.setString(1, id);
+            result = statementToGetLikes.executeQuery();
+            while (result.next()) {
+                post.getLikes().add(result.getString("username"));
+            }
+
+            PreparedStatement statementToGetHashTags = conn.prepareStatement(
+                    "select hashTag from hashtags where idOfPost = ?");
+            statementToGetHashTags.setString(1, id);
+            result = statementToGetHashTags.executeQuery();
+            while (result.next()) {
+                post.getHashTags().add(result.getString("hashTag"));
+            }
+            return post;
+
+        } catch(SQLException e) {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
+    public static String getLastId(DataSource dataSource) {
+        try (Connection conn = dataSource.getConnection()) {
+            ResultSet result = conn.prepareStatement("SELECT * FROM posts ORDER BY idOfPost DESC LIMIT 1;")
+                    .executeQuery();
+            result.next();
+            return result.getString(1);
+
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+        return "0";
+    }
+
+    public static void addPost(Post post, DataSource dataSource) {
+        int id = Integer.parseInt(getLastId(dataSource));
+        post.setId(id + 1 + "");
+        try (Connection conn = dataSource.getConnection()) {
+
+            PreparedStatement statementToInsertPost = conn.prepareStatement(
```

Comment:
Are you closing `PreparedStatement`s anywhere? They ought to be within the same `try(..)` as the connection. Otherwise it is the memory leak